grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java

```diff
@@ -374,11 +380,39 @@ private void doClose(Status status, Metadata metadata) {
         checkState(!closeCalled, "call already closed");
         closeCalled = true;
 
-        final HttpHeaders trailers = statusToTrailers(
-                ctx, sendHeadersCalled ? HttpHeaders.builder() : defaultHeaders.toBuilder(),
-                status, metadata);
+        final boolean trailersOnly;
+        if (firstResponse != null) {
+            // ResponseHeaders was written successfully.
+            trailersOnly = false;
+        } else {
+            if (!status.isOk() || responseHeaders == null) {
+                // Trailers-Only is permitted for calls that produce an immediate error.
+                // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+                trailersOnly = true;
+            } else {
+                if (method.getType().serverSendsOneMessage()) {
+                    // Incomplete request.
+                    trailersOnly = true;
```

Comment:
So we don't use `responseHeaders` in this case.
Is that correct?

Reply:
This is an error case. So I think we can produce a trailer-only response.